src/query/functions/src/aggregates/aggregate_histogram.rs

```diff
@@ -0,0 +1,615 @@
+    let mut lower_bucket_values = 0;
+
+    // Perform a binary search to find the maximum number of values that can fit into each bucket
+    let mut search_step = 0;
+    let max_search_steps = 10; // Limit the number of search steps to avoid excessive iteration
+
+    while upper_bucket_values > lower_bucket_values + 1 && search_step < max_search_steps {
+        // Calculate the midpoint of the upper and lower bounds
+        let bucket_values = (upper_bucket_values + lower_bucket_values) / 2;
+
+        // Check if the given number of values can be assigned to the desired number of buckets
+        if can_assign_into_buckets(value_map, bucket_values, num_buckets) {
+            // If it can, then set the upper bound to the midpoint
+            upper_bucket_values = bucket_values;
+        } else {
+            // If it can't, then set the lower bound to the midpoint
+            lower_bucket_values = bucket_values;
+        }
+        // Increment the search step counter
+        search_step += 1;
+    }
+
+    upper_bucket_values
+}
+
+/// ported from doris: https://github.com/apache/doris/blob/a1114d46e8c3f375325c176b602039987d8dea7b/be/src/vec/utils/histogram_helpers.hpp
+///
+/// Greedy equi-height histogram construction algorithm, inspired by the MySQL
+/// equi_height implementation(https://dev.mysql.com/doc/dev/mysql-server/latest/equi__height_8h.html).
+///
+/// Given an ordered collection of [value, count] pairs and a maximum bucket
+/// size, construct a histogram by inserting values into a bucket while keeping
+/// track of its size. If the insertion of a value into a non-empty bucket
+/// causes the bucket to exceed the maximum size, create a new empty bucket and
+/// continue.
+///
+/// The algorithm guarantees a selectivity estimation error of at most ~2 *
+/// #values / #buckets, often less. Values with a higher relative frequency are
+/// guaranteed to be placed in singleton buckets.
+///
+/// The minimum composite bucket size is used to minimize the worst case
+/// selectivity estimation error. In general, the algorithm will adapt to the
+/// data distribution to minimize the size of composite buckets. The heavy values
+/// can be placed in singleton buckets and the remaining values will be evenly
+/// spread across the remaining buckets, leading to a lower composite bucket size.
+///
+/// Note: The term "value" refers to an entry in a column and the actual value
+/// of an entry. The value_map is an ordered collection of [distinct value,
+/// value count] pairs. For example, a Value_map<String> could contain the pairs ["a", 1], ["b", 2]
+/// to represent one "a" value and two "b" values.
+///
+/// `value_map` An ordered map of distinct values and their counts.
+/// `max_num_buckets` The maximum number of buckets that can be used.
+///
+/// `return` the histogram buckets.
+fn build_histogram<T: Ord + Clone>(
+    value_map: &BTreeMap<T, u64>,
+    max_num_buckets: u64,
+) -> Vec<Bucket<T>> {
```

Comment:
You can use:

```
fn build_histogram<T: ValueType, S: Ord + Clone>(
    value_map: &BTreeMap<S, u64>,
    max_num_buckets: u64,
) -> Vec<Bucket<Scalar>>  {
   
   
   let v = T::upcast_scalar(s);
   ...
   
   // Scalar implements Display, so it would be easy to have pretty display result
  
 
}
```

Reply:
thx,at present, the format problem of `Date` and `DateTime` has been solved, but `scale` and `precision` of `Decimal` are placed on `type`, so I dealt with `Decimal` specially.